aten/src/ATen/native/Blas.cpp

```diff
@@ -4,60 +4,35 @@
-
-template<typename scalar_t>
-scalar_t dot_impl(int64_t n, scalar_t *x, int64_t incx, scalar_t *y, int64_t incy);
-
-template<typename scalar_t>
-scalar_t vdot_impl(int64_t n, scalar_t *x, int64_t incx, scalar_t *y, int64_t incy);
-
-constexpr inline bool lda_cond(int64_t m, int64_t n, int64_t lda) {
-  return n == 1 || lda >= std::max<int64_t>(1L, m);
-}
-
-Tensor &addmv_impl_cpu(Tensor& result, const Tensor &self, const Tensor &mat, const Tensor &vec, const Scalar& beta_, const Scalar& alpha_) {
-  auto r_stride = result.stride(0);
-  AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND(kBFloat16, mat.scalar_type(), "addmv_impl_cpu", [&] {
-    auto beta = beta_.to<scalar_t>();
-    auto alpha = alpha_.to<scalar_t>();
-    if (mat.stride(0) == 1 && lda_cond(mat.size(0), mat.size(1), mat.stride(1))) {
-      gemv<scalar_t>('n', mat.size(0), mat.size(1), alpha, mat.data_ptr<scalar_t>(), mat.stride(1),
-          vec.data_ptr<scalar_t>(), vec.stride(0), beta, result.data_ptr<scalar_t>(), r_stride);
-    }
-    else if (mat.stride(1) == 1 && lda_cond(mat.size(1), mat.size(0), mat.stride(0))) {
-      gemv<scalar_t>('t', mat.size(1), mat.size(0), alpha, mat.data_ptr<scalar_t>(), mat.stride(0),
-          vec.data_ptr<scalar_t>(), vec.stride(0), beta, result.data_ptr<scalar_t>(), r_stride);
-    }
-    else {
-      Tensor cmat = mat.contiguous();
-      gemv<scalar_t>('t', mat.size(1), mat.size(0), alpha, cmat.data_ptr<scalar_t>(), cmat.stride(0),
-          vec.data_ptr<scalar_t>(), vec.stride(0), beta, result.data_ptr<scalar_t>(), r_stride);
-    }
-  });
-  return result;
-}
-
-Tensor &addmv_out(const Tensor &self, const Tensor &mat, const Tensor &vec, const Scalar& beta, const Scalar& alpha, Tensor& result) {
-  { // scope of NoNamesGuard
-
-  at::NoNamesGuard guard;
-  result.resize_({mat.size(0)});
+namespace at {
+namespace meta {
+  TORCH_META_FUNC(addmv)(const Tensor &self, const Tensor &mat, const Tensor &vec, const Scalar& beta, const Scalar& alpha) {
 
   Tensor self_ = self;
-  if (self.dim() == 0 || self.size(0) == 1) {
+  if (self.numel()==1 && self.dim()<=1) {
     self_ = self.expand({mat.size(0)});
   }
+  //if self was broadcastable, it's now correct size
+  //if it wasn't, it'll error out in the subsequent checks
 
   TORCH_CHECK((mat.dim() == 2 && vec.dim() == 1 && self_.dim() == 1),
-    "vector + matrix @ vector expected, got ", self_.dim(), ", ", mat.dim(), ", ", vec.dim());
-  TORCH_CHECK((mat.size(1) == vec.size(0) && mat.size(0) == self_.size(0)),
-    "size mismatch, get ", self_.size(0), ", ", mat.size(0), "x", mat.size(1), ",", vec.size(0));
+    "vector + matrix @ vector expected, got ", self.dim(), ", ", mat.dim(), ", ", vec.dim());
 
-  if (mat.numel() == 0) {
+  TORCH_CHECK((mat.size(1) == vec.size(0) && mat.size(0) == self_.size(0)),
+    "size mismatch, got ", self_.size(0), ", ", mat.size(0), "x", mat.size(1), ",", vec.size(0));
+
+  std::array<int64_t, 1> out_size{mat.sizes()[0]};
+  set_output(out_size, mat.options());
```

Comment:
You don't need to allocate a `std::array` here, just pass in an initializer list `{ mat.sizes()[0] }` and that's sufficient

Reply:
Tried that, was a compile error (unless I messed something else up).

Reply:
Yep, that works, my initial attempt was busted because of wrong arguments to long form of `set_output` (which I'll still have to do for names).

Reply:
nit, and I haven't checked assembly so caveat emptor: Those are both probably less efficient than `IntArrayRef(mat.sizes().data(), 1)`, which shouldn't copy `mat.sizes()[0]` to the stack and take its address